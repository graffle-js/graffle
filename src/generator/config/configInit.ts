import type { IntrospectionOptions } from 'graphql'
import type { Fs } from '../../lib/fsp.js'
import type { Grafaid } from '../../lib/grafaid/__.js'
import type { Extension } from '../extension/types.js'

export interface InputLint {
  /**
   * Should a warning be raised if the GraphQL schema has a custom scalar defined for which your project has no corresponding custom scalar codec?
   *
   * @defaultValue `true`
   */
  missingCustomScalarCodec?: boolean
}

export const OutputCase = {
  pascal: `pascal`,
  camel: `camel`,
  kebab: `kebab`,
  snake: `snake`,
} as const
export type InputOutputCase = keyof typeof OutputCase

export const ImportFormat = {
  jsExtension: `jsExtension`,
  tsExtension: `tsExtension`,
  noExtension: `noExtension`,
} as const
export type InputImportFormat = keyof typeof ImportFormat

export interface ConfigInitSchemaSdl {
  type: `sdl`
  sdl: string
}
export interface ConfigInitSchemaInstance {
  type: `instance`
  instance: Grafaid.Schema.Schema
}
export interface ConfigInitSchemaSdlFile {
  type: `sdlFile`
  /**
   * Defaults to the source directory if set, otherwise the current working directory.
   */
  dirOrFilePath?: string
}
export interface ConfigInitSchemaUrl {
  type: `url`
  url: URL
  options?: InputIntrospectionOptions
}

export type ConfigInitSchema =
  | ConfigInitSchemaSdl
  | ConfigInitSchemaInstance
  | ConfigInitSchemaSdlFile
  | ConfigInitSchemaUrl
export interface ConfigInit {
  /**
   * File system API to use.
   *
   * By default uses the Node.js file system API.
   */
  fs?: Fs
  /**
   * What naming convention to use for generated files and directories.
   *
   * @defaultValue `kebab`
   */
  outputCase?: InputOutputCase
  /**
   * The name of the client. This will affect:
   *
   * 1. The field name the generated client is referenced under in the global registry.
   * 2. The exported namespace name.
   * 3. Possible use by extensions. For example the Opentelemetry extension augments span attributes with this value.
   *
   * By default the globally registered name is "default" and the exported namespace name is "Graffle".
   *
   * Tip: Typically you would only set this if you had multiple Graffle clients (for different GraphQL schemas) in your project.
   */
  name?: string
  /**
   * The namespace name to use.
   *
   * - If `true` then match the `name` if given.
   * - If `false` then use the default value.
   * - If a string then use that value.
   *
   * @defaultValue `Graffle`
   */
  nameNamespace?: boolean | string
  /**
   * Control various checks that Graffle performs. For example you could disable the warning about missing custom scalar codecs.
   */
  lint?: InputLint
  /**
   * The directory path which paths will be considered relative to.
   *
   * By default, is the process current working directory.
   */
  currentWorkingDirectory?: string
  /**
   * The schema to use for generation. Can be one of:
   *
   * 1. An existing SDL file on disk,
   * 2. A schema instance already in memory,
   * 3. An endpoint that will be introspected.
   */
  schema: ConfigInitSchema
  /**
   * If the schema comes from a non-sdl-file source like a GraphQL endpoint URL, should a derived SDL file be written to disk?
   *
   * When `boolean`:
   *
   * If true, an SDL file will be written into the output directory.
   *
   * When `string`:
   *
   * The path to write the SDL file to.
   * If a directory, then a file called "schema.graphql" will be written into it.
   * If a file, then it will be written as such.
   *
   * @defaultValue `false`
   */
  outputSDL?: boolean | string
  /**
   * Directory path to where the generated code should be output.
   *
   * Defaults to the current working directory.
   */
  outputDirPath?: string
  /**
   * Control over the client configuration's default schema. Since an introspection URL can be used for `schema`,
   * this option allows you to have this URL propagated to the generated client configuration for your convenience.
   * As well, its possible to set an explicit URL.
   *
   * When `boolean`:
   *
   * If a GraphQL endpoint is given for `schema`, should it be set as the default URL in the generated client.
   * If true then the client will default to using this URL when sending requests.
   *
   * When `URL`:
   *
   * A GraphQL endpoint to be used as the default URL in the generated client for requests.
   *
   * @defaultValue `true`
   */
  defaultSchemaUrl?: boolean | URL
  /**
   * Directory to look for custom scalar codecs.
   *
   * Defaults to the current working directory.
   */
  sourceDirPath?: string
  /**
   * File path to your scalars module.
   *
   * If not set, Graffle will look for a file called `scalars.ts` in the project directory.
   */
  scalars?: string
  /**
   * How should import identifiers be generated? Can be one of:
   *
   * 1. `jsExtension` e.g. `import ... from './bar.js'`
   * 2. `tsExtension` e.g. `import ... from './bar.ts'`
   * 3. `noExtension` e.g. `import ... from './bar'`
   *
   * @defaultValue `jsExtension`
   *
   * @remarks
   *
   * A user request for this option can be found at https://github.com/graffle-js/graffle/issues/1282.
   *
   * There is a planned feature to have a default be dynamic according to the state of your project's tsconfig.json.
   * See https://github.com/graffle-js/graffle/issues/1283.
   */
  importFormat?: InputImportFormat
  /**
   * Override import paths to graffle package within the generated code.
   * Used by Graffle test suite to have generated clients point to source
   * code. Probably not useful to you.
   */
  libraryPaths?: ConfigInitLibraryPaths
  /**
   * Should custom scalars definitions be imported into the generated output?
   */
  customScalars?: boolean
  /**
   * Should the generated code be formatted?
   */
  format?: boolean
  /**
   * Control over how TSDoc (JSDoc) comments are handled.
   */
  TSDoc?: {
    /**
     * When there is no documentation for a documentable node (type, field, enum value, ...) in the GraphQL schema
     * what should Graffle show for the JSDoc of the corresponding elements in the generated code?
     *
     * - `message`: Show a message mentioning the lack of documentation and the opportunity for it.
     * - `ignore`: Ignore the lack of documentation and show nothing.
     */
    noDocPolicy?: 'message' | 'ignore'
  }
  /**
   * Graffle gentime extensions to use. Most Graffle extensions are runtime only but some
   * are or have gentime components.
   */
  extensions?: Extension[]
}

export interface InputIntrospectionOptions extends IntrospectionOptions {
  /**
   * Whether to include descriptions in the introspection result.
   * @defaultValue `true`
   */
  descriptions?: boolean
  /**
   * Whether to include `specifiedByURL` in the introspection result.
   * @defaultValue `true`
   */
  specifiedByUrl?: boolean
  /**
   * Whether to include `isRepeatable` flag on directives.
   * @defaultValue `true`
   */
  directiveIsRepeatable?: boolean
  /**
   * Whether to include `description` field on schema.
   * @defaultValue `true`
   */
  schemaDescription?: boolean
  /**
   * Whether target GraphQL server support deprecation of input values.
   *
   * By default an attempt will be made to introspect this information
   * and if it fails then fallback to false for this option.
   */
  inputValueDeprecation?: boolean
  /**
   * Whether target GraphQL server supports `@oneOf` input objects.
   *
   * By default an attempt will be made to introspect this information
   * and if it fails then fallback to false for this option.
   */
  oneOf?: boolean
}

export interface ConfigInitLibraryPaths {
  client?: string
  schema?: string
  scalars?: string
  utilitiesForGenerated: string
  extensionTransportHttp: string
}

export type LibraryPathsKeys = keyof ConfigInitLibraryPaths

export const libraryPathKeys = {
  client: `client`,
  scalars: `scalars`,
  schema: `schema`,
  utilitiesForGenerated: `utilitiesForGenerated`,
  extensionTransportHttp: `extensionTransportHttp`,
} satisfies Record<LibraryPathsKeys, LibraryPathsKeys>
