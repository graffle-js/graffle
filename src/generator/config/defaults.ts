import type { ConfigInitLibraryPaths, InputImportFormat, InputOutputCase } from './configInit.js'

export const defaultNamespace = `Graffle`

export const defaultName = `default`

export const defaultLibraryPaths = {
  client: `graffle/client`,
  scalars: `graffle/generator-helpers/standard-scalar-types`,
  schema: `graffle/schema`,
  utilitiesForGenerated: `graffle/utilities-for-generated`,
  extensionTransportHttp: `graffle/extensions/transport-http`,
} satisfies ConfigInitLibraryPaths

export const defaultOutputCase: InputOutputCase = `kebab`

export const defaultImportFormat: InputImportFormat = `jsExtension`
