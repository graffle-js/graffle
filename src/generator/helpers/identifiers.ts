export const identifiers = {
  $$Schema: `$$Schema`,
  $$Scalar: `$$Scalar`,
  $$SelectionSets: `$$SelectionSets`,
  $$MethodsSelect: `$$MethodsSelect`,
  $$MethodsDocument: `$$MethodsDocument`,
  $$MethodsRoot: `$$MethodsRoot`,
  $$Data: `$$Data`,
  $$SchemaDrivenDataMap: `$$SchemaDrivenDataMap`,
  $$Utilities: `$$Utilities`,
  Schema: `Schema`,
  CustomScalars: `CustomScalars`,
  StandardScalar: `StandardScalar`,
}
