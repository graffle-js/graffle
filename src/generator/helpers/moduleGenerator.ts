import { camelCase, kebabCase, pascalCase, snakeCase } from 'es-toolkit'
import { casesExhausted } from '../../lib/prelude.js'
import type { Config } from '../config/config.js'
import {
  createCodeGenerator,
  type ModuleGeneratorRunner,
  type ModuleGeneratorRunnerImplementation,
} from './moduleGeneratorRunner.js'

export type FactoryModuleGenerator = <$Name extends string>(
  /**
   * The name of the file that will be written to disk.
   */
  name: $Name,
  runnerImplementation: ModuleGeneratorRunnerImplementation,
) => ModuleGenerator<$Name>

export interface ModuleGenerator<$Name extends string = string> {
  /**
   * The name of the file that will be written to disk.
   */
  name: $Name
  generate: ModuleGeneratorRunner
}

export interface GeneratedModule<$Name extends string = string> {
  name: $Name
  content: string
}

export const createModuleGenerator: FactoryModuleGenerator = (name, runnerImplementation) => {
  const runner = createCodeGenerator(runnerImplementation)

  const generate: ModuleGeneratorRunner = (config) => {
    const content = runner({ config })
    return {
      content,
      name,
    }
  }

  return {
    name,
    generate,
  }
}

export const importModuleGenerator = (config: Config, generator: ModuleGenerator) => {
  return `import * as $$${pascalCase(generator.name)} from './${getImportName(config, generator)}'`
}

export const getBaseName = (config: Config, generator: ModuleGenerator | GeneratedModule) => {
  return isExportsModule(generator.name)
    ? generator.name
    : caseFormatters[config.outputCase](generator.name)
}

export const getFileName = (config: Config, generator: ModuleGenerator | GeneratedModule) => {
  const name = getBaseName(config, generator)
  return `${name}.ts`
}

export const getImportName = (config: Config, generator: ModuleGenerator | GeneratedModule) => {
  const name = getBaseName(config, generator)
  switch (config.importFormat) {
    case `jsExtension`:
      return `${name}.js`
    case `tsExtension`:
      return `${name}.ts`
    case `noExtension`:
      return name
    default:
      throw casesExhausted(config.importFormat)
  }
}

export const caseFormatters = {
  pascal: pascalCase,
  camel: camelCase,
  kebab: kebabCase,
  snake: snakeCase,
}

export const isExportsModule = (name: string) => name.match(/^_+$/) !== null
