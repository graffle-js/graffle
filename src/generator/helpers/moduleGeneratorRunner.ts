import { CodePusher } from '#src/lib/code-pusher/index.js'
import type { Config } from '../config/config.js'
import type { GeneratedModule } from './moduleGenerator.js'

type FactoryModuleGeneratorRunner = <$CustomInput extends object = {}>(
  runnerImplementation: ModuleGeneratorRunnerImplementation<$CustomInput>,
) => CodeGenerator<$CustomInput>

export type ModuleGeneratorRunner = (config: Config) => GeneratedModule | GeneratedModule[]

export type CodeGenerator<$CustomInput extends object = {}> = (input: $CustomInput & BaseInput) => CodePusher.Code

export type ModuleGeneratorRunnerImplementation<$CustomInput extends object = {}> = (
  input: $CustomInput & BaseInputInternal,
) => void

interface BaseInputInternal extends BaseInput {
  code: CodePusher.CodePusher
}

interface BaseInput {
  config: Config
}

export const createCodeGenerator: FactoryModuleGeneratorRunner = (runnerImplementation) => {
  return (input) => {
    const codePusher = CodePusher.create()
    runnerImplementation({ ...input, code: codePusher })
    return codePusher.toString()
  }
}
