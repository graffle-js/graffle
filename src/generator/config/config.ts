import { pascalCase } from 'es-toolkit'
import * as Path from 'node:path'
import { Graffle } from '../../entrypoints/__Graffle.js'
import { Introspection } from '../../extensions/Introspection/Introspection.js'
import { ConfigManager } from '../../lib/config-manager/__.js'
import { fileExists, type Fs, isPathToADirectory, toAbsolutePath, toFilePath } from '../../lib/fsp.js'
import { Grafaid } from '../../lib/grafaid/__.js'
import { isString, keysStrict } from '../../lib/prelude.js'
import { type Formatter, getTypeScriptFormatter, passthroughFormatter } from '../../lib/typescript-formatter.js'
import type { Extension } from '../extension/types.js'
import {
  type ConfigInit,
  type ConfigInitLibraryPaths,
  type InputImportFormat,
  type InputLint,
  type InputOutputCase,
  libraryPathKeys,
} from './configInit.js'
import { defaultImportFormat, defaultLibraryPaths, defaultNamespace, defaultOutputCase } from './defaults.js'
import { defaultName } from './defaults.js'

export interface Config {
  fs: Fs
  name: string
  nameNamespace: string
  outputCase: InputOutputCase
  lint: Required<InputLint>
  schema: ConfigSchema
  runtimeFeatures: {
    customScalars: boolean
    operationVariables: boolean
  }
  options: {
    /**
     * Does the generated client import custom scalars statically from the user's project?
     */
    isImportsCustomScalars: boolean
    defaultSchemaUrl: URL | null
    format: boolean
    customScalars: boolean
    TSDoc: {
      noDocPolicy: 'message' | 'ignore'
    }
  }
  formatter: Formatter
  extensions: Extension[]
  importFormat: InputImportFormat
  paths: {
    project: {
      inputs: {
        root: string
        schema: null | string
        scalars: string
      }
      outputs: {
        sdl: null | string
        root: string
        modules: string
      }
    }
    imports: {
      scalars: string
      grafflePackage: Required<ConfigInitLibraryPaths>
    }
  }
}

interface ConfigSchema {
  via: ConfigInit['schema']['type']
  sdl: string
  sdlFilePath: null | string
  instance: Grafaid.Schema.Schema
  kindMap: Grafaid.Schema.KindMap
}

export const createConfig = async (configInit: ConfigInit): Promise<Config> => {
  // --- Fs ---

  const fs = configInit.fs ?? await import(`node:fs/promises`)

  // --- Output Case ---

  const outputCase = configInit.outputCase ?? defaultOutputCase

  // --- Paths ---

  const cwd = configInit.currentWorkingDirectory ?? process.cwd()

  const sourceDirPath = configInit.sourceDirPath ? toAbsolutePath(cwd, configInit.sourceDirPath) : cwd

  const outputDirPathRoot = configInit.outputDirPath
    ? toAbsolutePath(cwd, configInit.outputDirPath)
    : Path.join(cwd, `./graffle`)

  const outputDirPathModules = Path.join(outputDirPathRoot, `/modules`)

  const inputPathScalars = configInit.scalars
    ? toAbsolutePath(cwd, configInit.scalars)
    : Path.join(sourceDirPath, `scalars` + `.ts`)

  const isCustomScalarsModuleExists = await fileExists(fs, inputPathScalars)
  if (!isCustomScalarsModuleExists && configInit.scalars) {
    // dprint-ignore
    throw new Error(
      `Custom scalar codecs file not found. Given path: ${String(configInit.scalars)}. Resolved to and looked at: ${inputPathScalars}`,
    )
  }

  const scalarsImportPath = Path.relative(
    outputDirPathModules,
    inputPathScalars.replace(/\.ts$/, `.js`),
  )

  // --- Schema ---

  const schema = await createConfigSchema(fs, cwd, sourceDirPath, configInit)

  // --- Default Schema URL ---

  // dprint-ignore
  const defaultSchemaUrl =
    typeof configInit.defaultSchemaUrl === `boolean`
      ? configInit.schema instanceof Grafaid.Schema.Schema
        ? null
        : configInit.schema.type === `url`
          ? configInit.schema.url
          : null
      : configInit.defaultSchemaUrl ?? null

  // --- Formatting ---

  const formattingEnabled = configInit.format ?? true
  let formatter = passthroughFormatter
  if (formattingEnabled) {
    const formatterReal = await getTypeScriptFormatter(fs)
    if (!formatterReal) {
      // todo use floggy
      console.log(`
WARNING: No TypeScript formatter found. Generated code will remain ugly. To have code automatically formatted do one of the following things:

- pnpm add --save-dev @dprint/formatter @dprint/typescript
- pnpm add --save-dev prettier

To suppress this warning disable formatting in one of the following ways:

- CLI: graffle --no-format
- Configuration file: Generator.configuration({ format: false })
- API: Generator.generate({ format: false })
`.trim())
    } else {
      formatter = formatterReal
    }
  }

  // --- Library Paths ---

  const processLibraryPath = (path: string) => {
    const pathAbsolute = toAbsolutePath(cwd, path).replace(/\.ts$/, `.js`)
    return Path.relative(outputDirPathModules, pathAbsolute)
  }

  const libraryPaths = Object.fromEntries(
    keysStrict(libraryPathKeys).map(_ => {
      return [
        _,
        configInit.libraryPaths?.[_]
          ? processLibraryPath(configInit.libraryPaths[_])
          : undefined,
      ]
    }),
  )

  // --- Lint ---

  const lint: Config['lint'] = {
    missingCustomScalarCodec: configInit.lint?.missingCustomScalarCodec ?? true,
  }

  // --- Output SDL ---

  // dprint-ignore
  const outputSDLPath =
    configInit.outputSDL
      ? isString(configInit.outputSDL)
        ? toFilePath(`schema.graphql`, toAbsolutePath(cwd, configInit.outputSDL))
        : Path.join(outputDirPathRoot, `schema.graphql`)
      : null

  // --- name ---

  const name = configInit.name ?? defaultName

  const nameNamespace = configInit.nameNamespace === true
    ? configInit.name
      ? pascalCase(configInit.name)
      : defaultNamespace
    : isString(configInit.nameNamespace)
    ? configInit.nameNamespace
    : defaultNamespace

  // --- Config ---

  return {
    fs,
    name,
    importFormat: configInit.importFormat ?? defaultImportFormat,
    nameNamespace,
    extensions: configInit.extensions ?? [],
    outputCase,
    lint,
    formatter,
    runtimeFeatures: {
      customScalars: true, // todo do not assume true
      operationVariables: true, // todo do not assume true
    },
    schema,
    options: {
      isImportsCustomScalars: isCustomScalarsModuleExists,
      defaultSchemaUrl,
      format: formattingEnabled,
      customScalars: isCustomScalarsModuleExists,
      TSDoc: {
        noDocPolicy: configInit.TSDoc?.noDocPolicy ?? `ignore`,
      },
    },
    paths: {
      project: {
        outputs: {
          root: outputDirPathRoot,
          sdl: outputSDLPath,
          modules: outputDirPathModules,
        },
        inputs: {
          root: sourceDirPath,
          schema: schema.sdlFilePath,
          scalars: inputPathScalars,
        },
      },
      imports: {
        scalars: scalarsImportPath,
        grafflePackage: ConfigManager.mergeDefaults(
          defaultLibraryPaths,
          libraryPaths,
        ),
      },
    },
  }
}

const defaultSchemaFileName = `schema.graphql`

const createConfigSchema = async (
  fs: Fs,
  cwd: string,
  sourceDirPath: string,
  input: ConfigInit,
): Promise<ConfigSchema> => {
  switch (input.schema.type) {
    case `instance`: {
      const sdl = Grafaid.Schema.print(input.schema.instance)
      const instance = input.schema.instance
      const kindMap = Grafaid.Schema.KindMap.getKindMap(instance)
      return {
        via: input.schema.type,
        sdlFilePath: null,
        sdl,
        instance,
        kindMap,
      }
    }
    case `sdl`:
    case `sdlFile`: {
      let sdl
      let sdlFilePath: null | string = null
      if (input.schema.type === `sdlFile`) {
        const fileOrDirPath = input.schema.dirOrFilePath
          ? toAbsolutePath(cwd, input.schema.dirOrFilePath)
          : sourceDirPath
        const isDir = await isPathToADirectory(fs, fileOrDirPath)
        sdlFilePath = isDir ? Path.join(fileOrDirPath, defaultSchemaFileName) : fileOrDirPath
        sdl = await fs.readFile(sdlFilePath, `utf8`)
      } else {
        sdl = input.schema.sdl
      }
      const instance = Grafaid.Schema.buildSchema(sdl)
      const kindMap = Grafaid.Schema.KindMap.getKindMap(instance)
      return {
        via: input.schema.type,
        sdlFilePath,
        sdl,
        instance,
        kindMap,
      }
    }
    case `url`: {
      const introspection = Introspection({ options: input.schema.options })
      introspection
      const graffle = Graffle
        .create()
        .use(introspection)
        .transport({ url: input.schema.url })
      // todo introspection method should not be available without a transport
      const data = await graffle.introspect()
      if (!data) {
        throw new Error(`No data returned for introspection query.`)
      }
      const instance = Grafaid.Schema.buildClientSchema(data)
      const sdl = Grafaid.Schema.print(instance)
      const kindMap = Grafaid.Schema.KindMap.getKindMap(instance)
      return {
        via: `url`,
        sdlFilePath: null,
        sdl,
        instance,
        kindMap,
      }
    }
  }
}
