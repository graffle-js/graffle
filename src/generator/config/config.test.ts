import { expect } from 'vitest'
import { describe } from 'vitest'
import { test } from '../../../tests/_/helpers.js'
import { createConfig } from './config.js'
import type { ConfigInitSchemaSdl } from './configInit.js'

const schema: ConfigInitSchemaSdl = {
  type: `sdl`,
  sdl: `type Query { ok: Boolean }`,
}

describe(`import format`, () => {
  test(`defaults to jsExtension`, async () => {
    const config = await createConfig({ schema })
    expect(config.importFormat).toEqual(`jsExtension`)
  })
  test(`noExtension`, async () => {
    const customPathFile = `./tests/_/fixtures/custom.graphql`
    const config = await createConfig({
      schema: { type: `sdlFile`, dirOrFilePath: customPathFile },
      importFormat: `noExtension`,
    })
    expect(config.importFormat).toEqual(`noExtension`)
  })
})

test(`can load schema from custom path`, async () => {
  const customPathFile = `./tests/_/fixtures/custom.graphql`
  const config = await createConfig({ schema: { type: `sdlFile`, dirOrFilePath: customPathFile } })
  const field = config.schema.instance.getQueryType()?.getFields()[`customNamedSchemaFile`]
  expect(config.paths.project.inputs.schema).match(new RegExp(customPathFile + `$`))
  expect(config.schema.sdl).toMatchSnapshot()
  expect(field).toBeDefined()
})

test(`can load schema from custom dir using default file name`, async () => {
  const customPathDir = `tests/_/fixtures`
  const config = await createConfig({ schema: { type: `sdlFile`, dirOrFilePath: customPathDir } })
  const field = config.schema.instance.getQueryType()?.getFields()[`defaultNamedSchemaFile`]
  expect(config.paths.project.inputs.schema).match(new RegExp(customPathDir + `/schema.graphql$`))
  expect(config.schema.sdl).toMatchSnapshot()
  expect(field).toBeDefined()
})

test(`can introspect schema from url`, async ({ pokemonService }) => {
  const config = await createConfig({ schema: { type: `url`, url: pokemonService.url } })
  expect(config.paths.project.inputs.schema).toEqual(null)
  expect(config.schema.sdl).toMatchSnapshot()
})

test(`configured schema introspection options are passed to introspection`, async ({ pokemonService, fetch }) => {
  fetch.mockImplementation(_ => {
    const response = new Response(JSON.stringify({ data: null }))
    return Promise.resolve(response)
  })
  await createConfig({
    schema: {
      type: `url`,
      url: pokemonService.url,
      options: {
        descriptions: false,
        directiveIsRepeatable: false,
        inputValueDeprecation: false,
        schemaDescription: false,
        oneOf: false,
        specifiedByUrl: false,
      },
    },
  }).catch((_: unknown) => {})
  const readableStream: ReadableStream = fetch.mock.calls[0]?.[0]?.body as any
  const { value }: { value: Uint8Array } = await readableStream.getReader().read() as any
  const document = JSON.parse(new TextDecoder().decode(value))
  expect(document).toMatchSnapshot()
})
