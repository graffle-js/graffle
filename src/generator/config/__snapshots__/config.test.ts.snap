// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can introspect schema from url 1`] = `
"union Battle = BattleRoyale | BattleTrainer | BattleWild

type BattleRoyale {
  combatants: [CombatantMultiPokemon!]
  date: Float
  id: ID
  winner: Trainer
}

type BattleTrainer {
  combatant1: CombatantSinglePokemon
  combatant2: CombatantSinglePokemon
  date: Float
  id: ID
  winner: Trainer
}

type BattleWild {
  date: Float
  id: ID
  pokemon: Pokemon
  result: BattleWildResult
  trainer: Trainer
  wildPokemons: [Pokemon!]
}

enum BattleWildResult {
  pokemonsCaptured
  pokemonsDefeated
  trainerDefeated
}

interface Being {
  id: ID
  name: String
}

type CombatantMultiPokemon {
  pokemons: [Pokemon!]
  trainer: Trainer
}

type CombatantSinglePokemon {
  pokemon: Pokemon
  trainer: Trainer
}

"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.This scalar is serialized to a string in ISO 8601 format and parsed from a string in ISO 8601 format.
"""
scalar Date

input DateFilter {
  gte: Date
  lte: Date
}

type Mutation {
  addPokemon(attack: Int, defense: Int, hp: Int, name: String!, type: PokemonType!): Pokemon
}

type Patron implements Being {
  id: ID
  money: Int
  name: String
}

type Pokemon implements Being {
  attack: Int
  birthday: Date
  defense: Int
  hp: Int
  id: ID
  name: String
  trainer: Trainer
  type: PokemonType
}

input PokemonFilter {
  birthday: DateFilter
  name: StringFilter
}

enum PokemonType {
  bug
  electric
  fire
  grass
  water
}

type Query {
  battles: [Battle!]!
  beings: [Being!]!
  pokemon: [Pokemon!]
  pokemonByName(name: String!): [Pokemon!]
  pokemons(filter: PokemonFilter): [Pokemon!]
  trainerByName(name: String!): Trainer
  trainers: [Trainer!]
}

input StringFilter {
  contains: String
  in: [String!]
}

type Trainer implements Being {
  class: TrainerClass
  fans: [Patron!]
  id: ID
  name: String
  pokemon: [Pokemon!]
}

enum TrainerClass {
  bugCatcher
  camper
  picnicker
  psychic
  psychicMedium
  psychicYoungster
  sailor
  superNerd
  tamer
  teamRocketGrunt
  triathlete
  youngster
  youth
}"
`;

exports[`can load schema from custom dir using default file name 1`] = `
"type Query {
	defaultNamedSchemaFile: Boolean
}
"
`;

exports[`can load schema from custom path 1`] = `
"type Query {
	customNamedSchemaFile: Boolean
}
"
`;

exports[`configured schema introspection options are passed to introspection 1`] = `
{
  "query": "
    query IntrospectionQuery {
      __schema {
        
        queryType { name }
        mutationType { name }
        subscriptionType { name }
        types {
          ...FullType
        }
        directives {
          name
          
          
          locations
          args {
            ...InputValue
          }
        }
      }
    }

    fragment FullType on __Type {
      kind
      name
      
      
      
      fields(includeDeprecated: true) {
        name
        
        args {
          ...InputValue
        }
        type {
          ...TypeRef
        }
        isDeprecated
        deprecationReason
      }
      inputFields {
        ...InputValue
      }
      interfaces {
        ...TypeRef
      }
      enumValues(includeDeprecated: true) {
        name
        
        isDeprecated
        deprecationReason
      }
      possibleTypes {
        ...TypeRef
      }
    }

    fragment InputValue on __InputValue {
      name
      
      type { ...TypeRef }
      defaultValue
      
      
    }

    fragment TypeRef on __Type {
      kind
      name
      ofType {
        kind
        name
        ofType {
          kind
          name
          ofType {
            kind
            name
            ofType {
              kind
              name
              ofType {
                kind
                name
                ofType {
                  kind
                  name
                  ofType {
                    kind
                    name
                    ofType {
                      kind
                      name
                      ofType {
                        kind
                        name
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  ",
}
`;
